File: attachement/data_processing.R

# library(assertr)
library(rhdf5)
library(Matrix) 


get_omic <- function(path) {
  path_parts <- unlist(strsplit(path, "/"))
  return(path_parts[length(path_parts) - 1])
}

omic2list_name = list(mixRNA = "mix_rna" , 
mixMET =  "mix_met",
MET='ref_met',
RNA= 'ref_bulkRNA',
scRNA= 'ref_scRNA' ) 


write_data_frame <- function(name, path,data){
    h5createGroup(path, name)
    h5write(data, file = path, name = paste0(name, "/data"))

    
    if(length(colnames(data))){
      h5write(colnames(data), file = path, name = paste0(name, "/samples"))
    }
    if(length(rownames(data))){
      h5write(rownames(data), file = path, name = paste0(name, "/genes"))
    }



}


write_sparse_matrix <- function(group, path, counts, meta, data = NULL, scale_data = NULL) {
  h5createGroup(path, group)

  # Write counts
  h5createDataset(path, paste0(group, '/data'), dims = length(counts@x), storage.mode = "double")
  h5write(counts@x, path, paste0(group, "/data"))
  h5write(dim(counts), path, paste0(group, "/shape"))
  h5createDataset(path, paste0(group, '/indices'), dims = length(counts@i), storage.mode = "integer")
  h5write(counts@i, path, paste0(group, "/indices"))
  h5write(counts@p, path, paste0(group, "/indptr"))
  if (!is.null(rownames(counts))) {
    h5write(rownames(counts), path, paste0(group, "/genes"))
  }
  if (!is.null(colnames(counts))) {
    h5write(colnames(counts), path, paste0(group, "/cell"))
  }

  h5write(meta, path, paste0(group, "/meta"))

  if (!is.null(data)) {
    h5createDataset(path, paste0(group, '/normalized_data'), dims = length(data@x), storage.mode = "double")
    h5write(data@x, path, paste0(group, "/normalized_data"))
  }

}


write_hdf5 <- function(path, data_list) {
  # Create the HDF5 file
  h5createFile(path)

  for (name in names(data_list)) {

    if (name == "ref_scRNA") {
      h5createGroup(path, 'ref_scRNA/')

      for (dataset in names(data_list[[name]])) {
        group = paste0('ref_scRNA/', dataset)
        h5createGroup(path, group)

        # entry <- data_list[[name]][[dataset]]
        seurat_obj <- NULL
        seurat_field_name <- NULL
        group_seurat = group
        if(inherits(data_list[[name]][[dataset]], "Seurat")){
          seurat_obj <- data_list[[name]][[dataset]]
          seurat_field_name <- dataset
          
        }

        ##Check if there is a Seurat Object 
        for (field in names(data_list[[name]][[dataset]])) {
          if (inherits(data_list[[name]][[dataset]][[field]], "Seurat")) {
            seurat_obj <- data_list[[name]][[dataset]][[field]]
            seurat_field_name <- field
            group_seurat = paste0('ref_scRNA/', dataset,'/',field)
            # h5createGroup(path,paste0(path,group), field)
            h5createGroup(path,group_seurat)

            break
          }
        }

        if (!is.null(seurat_obj)) {
          library(Seurat)

          h5write("seurat", path, paste0(group_seurat, "/object_type"))
          h5write(seurat_field_name, path, paste0(group_seurat, "/seurat_field_name"))

          write_sparse_matrix(group_seurat, path,
            counts = GetAssayData(seurat_obj, assay = "RNA", layer = "counts"),
            meta = seurat_obj@meta.data , 
            data = GetAssayData(seurat_obj, assay = "RNA", layer = "data"), 
            scale_data= GetAssayData(seurat_obj, assay = "RNA", layer = "scale.data")
          )

        }
      
        if("counts" %in% names(data_list[[name]][[dataset]]) && "metadata" %in% names(data_list[[name]][[dataset]])) {
          write_sparse_matrix(group, path, data_list[[name]][[dataset]]$counts, data_list[[name]][[dataset]]$metadata)
        }
      }
    } else {
      write_data_frame(name, path, data_list[[name]])
    }
  }
}



read_data_frame <- function(path,name,file_structure){
    data <- h5read(file = path, name = paste0("/",name, "/data"))
    
    if(!is.null(file_structure[[name]]$samples)){
      samples <- h5read(file = path, name = paste0("/",name, "/samples"))
      colnames(data) <- samples
    }
    if(!is.null(file_structure[[name]]$cell_types)){
      cell_types <- h5read(file = path, name = paste0("/",name, "/cell_types"))
      colnames(data) <- cell_types
    }
    if(!is.null(file_structure[[name]]$genes)){
      genes <- h5read(file = path, name = paste0("/",name, "/genes"))
      rownames(data) <- genes
    }
    if(!is.null(file_structure[[name]]$CpG_sites)){
      CpG_sites <- h5read(file = path, name = paste0("/",name, "/CpG_sites"))
      rownames(data) <- CpG_sites
    }
    return(data)

}

read_sparse_matrix <- function(group,group_structure, path) {
  counts_data <- as.numeric(h5read(path, paste0(group, "/data")))
  counts_shape <- as.integer(h5read(path, paste0(group, "/shape")))
  counts_indices <- as.integer(h5read(path, paste0(group, "/indices")))
  counts_indptr <- as.integer(h5read(path, paste0(group, "/indptr")))
  gene_names <- as.character(h5read(path, paste0(group, "/genes")))
  cell_names <- as.character(h5read(path, paste0(group, "/cell")))

  counts <- new("dgCMatrix",
                x = counts_data,
                i = counts_indices,
                p = counts_indptr,
                Dim = counts_shape,
                Dimnames = list(gene_names, cell_names))

  meta <- h5read(path, paste0(group, "/meta"))
  meta <- as.data.frame(meta, stringsAsFactors = FALSE)
  rownames(meta) <- cell_names

  # Detect if the original object was Seurat
  is_seurat <- FALSE
  # if (paste0(group, "/object_type") %in% h5ls(path, recursive = TRUE)$name) {
  if ( "object_type" %in% names(group_structure)) {
    object_type <- h5read(path, paste0(group, "/object_type"))
    is_seurat <- object_type == "seurat"
  }

  if (is_seurat) {
    library(Seurat)
    seurat_obj <- CreateSeuratObject(counts = counts, meta.data = meta)
    DefaultAssay(seurat_obj) <- "RNA"
    if ("normalized_data" %in% names(group_structure)) {
      data <- counts
      data@x <- as.numeric(h5read(path, paste0(group, "/normalized_data")))
      seurat_obj[["RNA"]] <- SetAssayData(seurat_obj[["RNA"]], layer = "data", new.data = data)
    }

    if ("scale_data" %in% names(group_structure)) {
      scale_values <- as.numeric(h5read(path, paste0(group, "/scale_data")))
      scale_data <- counts
      scale_data@x <- scale_values
      seurat_obj[["RNA"]] <- SetAssayData(seurat_obj[["RNA"]], layer = "scale.data", new.data = scaled_data)

    }

    return(seurat_obj)
  } else {
    return(list(counts = counts, metadata = meta))
  }
}

read_hdf5 <- function(path) {
  file_structure <- h5dump(path, load = FALSE)
  group_names <- names(file_structure)
  data_list <- list()

  for (name in group_names) {
    if (name == "ref_scRNA") {
      ref_scRNA <- list()

      for (dataset in names(file_structure[[name]])) {
        seuratobj_or_counts_n_metadata <- read_sparse_matrix(paste0('ref_scRNA/', dataset), file_structure[[name]][[dataset]] , path)
        
        ## check if there is another data inside here !  TODO change this with a metadata list in the root of this file... 
        expected_names <- c("cell", "data", "genes", "indices", "indptr", "meta", "shape")
        actual_names <- names(file_structure[[name]][[dataset]])
        unexpected_names = actual_names[!actual_names %in% expected_names]

        for(field in unexpected_names ){
          seuratobj_or_counts_n_metadata[[field]] = read_sparse_matrix(paste0('ref_scRNA/', dataset,'/',field), file_structure[[name]][[dataset]][[field]] , path)
        }
        ref_scRNA[[dataset]] = seuratobj_or_counts_n_metadata
      }

      data_list[[name]] <- ref_scRNA
    } else {
      data_list[[name]] <- read_data_frame(path, name, file_structure)
    }
  }

  return(data_list)
}

File: attachement/data_processing.py
import numpy as np
import h5py
from scipy.sparse import csc_matrix, issparse
import pandas as pd

compression_lvl = 6
compression_type = "gzip"


def read_all_ref_hdf5(path):
    with h5py.File(path, "r") as f:
        # Read ref_bulkRNA data
        ref_bulkRNA = pd.DataFrame(f["ref_bulkRNA/data"][:]).T
        ref_bulkRNA.columns = [x.decode('utf-8') for x in f["ref_bulkRNA/cell_types"][:]]
        ref_bulkRNA.index = [x.decode('utf-8') for x in f["ref_bulkRNA/genes"][:]]


        # Read ref_met data
        ref_met =  pd.DataFrame(f["ref_met/data"][:]).T
        ref_met.columns = [x.decode('utf-8') for x in f["ref_met/cell_types"][:]]
        ref_met.index = [x.decode('utf-8') for x in f["ref_met/CpG_sites"][:]]


        # Read ref_scRNA data
        ref_scRNA = {}
        datasets = ["ref_sc_peng", "ref_sc_baron", "ref_sc_raghavan"]

        for dataset in datasets:
            group = f"ref_scRNA/{dataset}"

            counts_data = f[f"{group}/data"][:]
            counts_shape = tuple(f[f"{group}/shape"][:])
            counts_indices = f[f"{group}/indices"][:]
            counts_indptr = f[f"{group}/indptr"][:]

            genes = [x.decode('utf-8') for x in f[f"{group}/genes"][:]] if f"{group}/genes" in f else None
            cells = [x.decode('utf-8') for x in f[f"{group}/cell"][:]] if f"{group}/cell" in f else None

            adata = ad.AnnData(csc_matrix((counts_data, counts_indices, counts_indptr), shape=counts_shape,dtype=np.int32).T)

            if genes is not None:
                adata.var_names = genes

            if cells is not None:
                adata.obs_names  = cells

            meta = pd.DataFrame(f[f"{group}/meta"][:])
            meta.index = cells
            for col in meta.columns:
                if meta[col].dtype == 'object':  # Check if the column contains byte strings
                    meta[col] = meta[col].apply(lambda x: x.decode('utf-8') if isinstance(x, bytes) else x)
            meta = meta.convert_dtypes()
            # meta = encode_strings(meta)

            ref_scRNA[dataset] = {
                "counts": adata,
                "metadata": meta
            }

        ref_all = {
            "ref_bulkRNA": ref_bulkRNA,
            "ref_met": ref_met,
            "ref_scRNA": ref_scRNA
        }

    return ref_all




def write_sparse_matrix(group, fw, counts, meta, data=None, scale_data=None,
                        compression_type="gzip", compression_lvl=6):

    # Write sparse matrix components
    group.create_dataset("data", data=counts.data, compression=compression_type, compression_opts=compression_lvl,chunks= (len(counts.data)),shuffle=True)
    group.create_dataset("shape",   data=counts.shape, compression=compression_type, compression_opts=compression_lvl,chunks=   (len(counts.shape)),shuffle=True)
    group.create_dataset("indices", data=counts.indices, compression=compression_type, compression_opts=compression_lvl,chunks= (len(counts.indices)),shuffle=True)
    group.create_dataset("indptr",  data=counts.indptr, compression=compression_type, compression_opts=compression_lvl,chunks=  (len(counts.indptr)),shuffle=True)

    if hasattr(counts, 'var_names'):
        group.create_dataset("genes", data=np.array(counts.var_names, dtype='S'), compression=compression_type,chunks=  (len(counts.var_names)),shuffle=True)
    if hasattr(counts, 'obs_names'):
        group.create_dataset("cell", data=np.array(counts.obs_names, dtype='S'), compression=compression_type,chunks= (len(counts.obs_names)),shuffle=True)

    # Write metadata
    dtype_list = []
    for col in meta.columns:
        if pd.api.types.is_integer_dtype(meta[col]):
            dtype_list.append((col, np.int32))
        else:
            dtype_list.append((col, h5py.string_dtype('utf-8')))
    meta_np = np.array(list(meta.itertuples(index=False, name=None)), dtype=dtype_list)
    group.create_dataset("meta", data=meta_np, compression=compression_type, compression_opts=compression_lvl,chunks= (len(meta_np)),shuffle=True)

    # Optionally write normalized data
    if data is not None and issparse(data):
        group.create_dataset("normalized_data", data=data.data, compression=compression_type, compression_opts=compression_lvl,chunks= (len(data.data)),shuffle=True)

    # Optionally write scale.data
    if scale_data is not None:
        if issparse(scale_data):
            group.create_dataset("scale_data", data=scale_data.data, compression=compression_type, compression_opts=compression_lvl,chunks= (len(scale_data.data)),shuffle=True)
        elif isinstance(scale_data, np.ndarray) or isinstance(scale_data, pd.DataFrame):
            scale_arr = np.array(scale_data, dtype=np.float32)
            group.create_dataset("scale_data", data=scale_arr.flatten(), compression=compression_type, compression_opts=compression_lvl,chunks= (len(scale_arr)),shuffle=True)
            group.create_dataset("scale_data_shape", data=scale_arr.shape, compression=compression_type, compression_opts=compression_lvl,chunks= (len(scale_arr.shape)),shuffle=True)
            if isinstance(scale_data, pd.DataFrame):
                group.create_dataset("scale_data_genes", data=np.array(scale_data.index, dtype='S'), compression=compression_type, compression_opts=compression_lvl,chunks= (len(scale_data.index)),shuffle=True)
                group.create_dataset("scale_data_cells", data=np.array(scale_data.columns, dtype='S'), compression=compression_type, compression_opts=compression_lvl,chunks= (len(scale_data.columns)),shuffle=True)


# def write_global_hdf5(path, data_list, compression_type="gzip", compression_lvl=4):
#     with h5py.File(path, "w") as fw:
#         for name in data_list:
#             if name == "ref_scRNA":
#                 group_ref = fw.create_group("ref_scRNA")

#                 for dataset in data_list[name]:
#                     group_path = f"ref_scRNA/{dataset}"
#                     group = fw.create_group(group_path)

#                     # Check if dataset is Seurat-like object (dict with laye        rs)
#                     if "counts" in data_list[name][dataset] and "metadata" in data_list[name][dataset]:
#                         # counts = csc_matrix(data_list[name][dataset]["counts"].X.T)
#                         counts = data_list[name][dataset]["counts"]
#                         meta = data_list[name][dataset]["metadata"]
#                         data = None
#                         scale_data = None

#                         if "data" in data_list[name][dataset]:
#                             # data = csc_matrix(data_list[name][dataset]["data"].X.T)
#                             data = data_list[name][dataset]["data"]
#                         if "scale.data" in data_list[name][dataset]:
#                             scale_data = data_list[name][dataset]["scale.data"]

#                         write_sparse_matrix(group, fw, counts, meta, data=data, scale_data=scale_data,
#                                             compression_type=compression_type, compression_lvl=compression_lvl)
                    
#                     else:
#                         # If not a Seurat object but still has "counts" and "metadata"
#                         if "counts" in data_list[name][dataset] and "metadata" in data_list[name][dataset]:
#                             counts = csc_matrix(data_list[name][dataset]["counts"].X.T)
#                             meta = data_list[name][dataset]["metadata"]
#                             write_sparse_matrix(group_path, fw, counts, meta, compression_type=compression_type, compression_lvl=compression_lvl)

#             else:
#                 df = data_list[name]
#                 df_group = fw.create_group(name)
#                 df_group.create_dataset("data",data=df.T, compression=compression_type, compression_opts=compression_lvl)
#                 df_group.create_dataset("samples",data=list(map(str,df.columns)), compression=compression_type, compression_opts=compression_lvl)
#                 df_group.create_dataset("genes",data=list(map(str,df.index)), compression=compression_type, compression_opts=compression_lvl)
#                 # for col in df.columns:
#                 #     df_group.create_dataset(col, data=np.array(df[col]), compression=compression_type, compression_opts=compression_lvl)


import h5py
import numpy as np
import pandas as pd
from scipy.sparse import csc_matrix

def write_hdf5(path, data_list, compression_type="gzip", compression_lvl=4):
    with h5py.File(path, "w") as fw:
        for name, value in data_list.items():
            if name == "ref_scRNA":
                group_ref = fw.create_group("ref_scRNA")

                for dataset, content in value.items():
                    group_path = f"ref_scRNA/{dataset}"
                    group = fw.create_group(group_path)

                    if isinstance(content, dict) and "counts" in content and "metadata" in content:
                        counts = content["counts"]
                        meta = content["metadata"]
                        data = content.get("data")
                        scale_data = content.get("scale.data")

                        write_sparse_matrix(
                            group, fw, counts, meta,
                            data=data, scale_data=scale_data,
                            compression_type=compression_type,
                            compression_lvl=compression_lvl
                        )

                    else:
                        # fallback if not Seurat-like
                        if isinstance(content, dict) and "counts" in content and "metadata" in content:
                            counts = csc_matrix(content["counts"].X.T)
                            meta = content["metadata"]
                            write_sparse_matrix(
                                group_path, fw, counts, meta,
                                compression_type=compression_type,
                                compression_lvl=compression_lvl
                            )

            # --- 2️ Normal DataFrame ---
            elif isinstance(value, pd.DataFrame):
                df_group = fw.create_group(name)
                df_group.create_dataset(
                    "data", data=value.T, 
                    compression=compression_type, compression_opts=compression_lvl
                )
                df_group.create_dataset(
                    "samples", data=[str(x) for x in value.columns],
                    compression=compression_type, compression_opts=compression_lvl
                )
                df_group.create_dataset(
                    "genes", data=[str(x) for x in value.index],
                    compression=compression_type, compression_opts=compression_lvl
                )

            # --- 3️ Simple numeric or scalar value ---
            elif np.isscalar(value):
                grp = fw.create_group(name)
                grp.create_dataset("data", data=float(value))
                continue

            # --- 4️ Numpy array or list ---
            elif isinstance(value, (np.ndarray, list)):
                fw.create_dataset(name, data=value, compression=compression_type, compression_opts=compression_lvl)

            # --- 5️ Unknown type fallback ---
            else:
                grp = fw.create_group(name)
                grp.create_dataset("data", data=np.string_(str(value)))
                # fw.create_dataset(name, data=np.string_(str(value)))


def set_dataframe_index_and_columns(group, group_name, df):
    def get_string_array(field_name):
        if field_name in group_name :
            raw = group[field_name][()]
            if raw.dtype.kind in {'S', 'O'}:
                return [x.decode('utf-8') if isinstance(x, bytes) else x for x in raw]
            return list(raw)
        return None

    # Assign columns
    for field in ["samples", "cell_types"]:
        values = get_string_array(field)
        if values is not None:
            df.columns = values
            break  # Prioritize 'samples' over 'cell_types' like in R

    # Assign rows
    for field in ["genes", "CpG_sites"]:
        values = get_string_array(field)
        if values is not None:
            df.index = values
            break  # Prioritize 'genes' over 'CpG_sites'
    return df

def get_h5_structure(h5group):
    structure = {}
    for key in h5group:
        if isinstance(h5group[key], h5py.Group):
            structure[key] = get_h5_structure(h5group[key])
        else:
            structure[key] = "dataset"
    return structure

# def read_data_frame(f, group_name, file_structure):
#     # with h5py.File(path, "r") as f:
#     group = f[group_name]
#     df = pd.DataFrame(group["data"][()]).T

#     df = set_dataframe_index_and_columns(group, file_structure[group_name],  df)

#     return df

def read_data_frame(f, group_name, file_structure):
    group = f[group_name]
    data = group["data"]

    # Check if this is a compound dataset
    is_compound = data.dtype.names is not None

    arr = data[()]
    if not is_compound and arr.ndim == 2:
        df = pd.DataFrame(arr.T)
    else:
        df = pd.DataFrame(arr)

    df = set_dataframe_index_and_columns(group, file_structure[group_name], df)
    return df


def read_sparse_matrix(group_path, group_structure, h5file):
    # Load sparse matrix components
    counts_data = np.array(h5file[f"{group_path}/data"])
    counts_shape = tuple(h5file[f"{group_path}/shape"])
    counts_indices = np.array(h5file[f"{group_path}/indices"])
    counts_indptr = np.array(h5file[f"{group_path}/indptr"])

    gene_names = [g.decode("utf-8") for g in h5file[f"{group_path}/genes"][:]]
    cell_names = [c.decode("utf-8") for c in h5file[f"{group_path}/cell"][:]]

    # Reconstruct sparse matrix
    counts = csc_matrix((counts_data, counts_indices, counts_indptr), shape=counts_shape)
    counts.var_names = gene_names
    counts.obs_names = cell_names

    # Load metadata
    meta_data_raw = h5file[f"{group_path}/meta"]
    meta = pd.DataFrame({key: meta_data_raw[key][:] for key in meta_data_raw.dtype.names})
    meta.columns = meta_data_raw.dtype.names
    meta.index = cell_names

    # Check if it was a Seurat object
    is_seurat = "object_type" in group_structure and h5file[f"{group_path}/object_type"][()].decode("utf-8") == "seurat"

    if is_seurat:
        # Add normalized data layer
        normalized_data = None
        if "normalized_data" in group_structure:
            normalized_data = csc_matrix((h5file[f"{group_path}/normalized_data"][:], counts_indices, counts_indptr), shape=counts_shape)

        # Add scaled data
        scale_data = None
        if "scale_data" in group_structure:
            scale_data = csc_matrix((h5file[f"{group_path}/scale_data"][:], counts_indices, counts_indptr), shape=counts_shape)

        return {
            "counts": counts,
            "metadata": meta,
            "data": normalized_data,
            "scale.data": scale_data,
            "is_seurat": True
        }

    else:
        return {
            "counts": counts,
            "metadata": meta
        }



def read_hdf5(path):
    data_list = {}

    with h5py.File(path, "r") as f:
        file_structure = get_h5_structure(f)

        for name in file_structure:
            if name == "ref_scRNA":
                ref_scRNA = {}
                for dataset in file_structure[name]:
                    group_path = f"{name}/{dataset}"
                    seurat_data = read_sparse_matrix(group_path, file_structure[name][dataset], f)

                    # Look for nested groups (fields like scaled_data, data, etc.)
                    expected = {"cell", "data", "genes", "indices", "indptr", "meta", "shape", "object_type", "seurat_field_name", "normalized_data", "scale_data"}
                    actual = set(file_structure[name][dataset].keys())
                    nested_fields = actual - expected

                    for field in nested_fields:
                        sub_group_path = f"{name}/{dataset}/{field}"
                        seurat_data[field] = read_sparse_matrix(sub_group_path, file_structure[name][dataset][field], f)

                    ref_scRNA[dataset] = seurat_data

                data_list[name] = ref_scRNA

            else:
                data_list[name] = read_data_frame(f, name, file_structure)


    return data_list

